### DISCUSSION TITLE
How to modify the response of OnRecordsListRequest event hook

### DISCUSSION THREAD

**Discussion by elmarw** [pocketbase/pocketbase#1805] [Q&A] [answered]
According to the docs you can modify the response with the following hook:

`
App.OnRecordsListRequest().Add(func(e *core.RecordsListEvent) error {
    return nil
})
`

Our use case is the following. We have a database with user reviews and we would like to modify the response so that it also shows the average rating of all the reviews.

on "core.RecordsListEvent" you can see the Result, but of course you can't modify the type. 

Is there any way to add data before its returned by echo's JSON function? What would be the recommended way? Might be nice to add a "metadata" any type to the search.Result struct?

**Comment by ganigeorgiev** [comment, marked as answer]
You should be able to write a custom json response if you return `hook.StopPropagation` (this in the future may not be necessary since I'm planning some changes in relation to the hooks_).

> Might be nice to add a "metadata" any type to the search.Result struct?

It is possible but I'm not sure how useful that would be since it is not that difficult to return a different response if you really want to, something like:

```js
package main

import (
	"log"
	"net/http"

	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tools/hook"
)

func main() {
	app := pocketbase.New()

	app.OnRecordsListRequest("reviews").Add(func(e *core.RecordsListEvent) error {
		data := map[string]any{
			"page":       e.Result.Page,
			"perPage":    e.Result.PerPage,
			"totalItems": e.Result.TotalItems,
			"totalPages": e.Result.TotalPages,
			"items":      e.Result.Items,
			"meta": map[string]any{
				// anything else
				"a": "test1",
				"b": 456,
			},
		}

		e.HttpContext.JSON(http.StatusOK, data)

		return hook.StopPropagation
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

**Reply by elmarw** [reply]
That's really great. Thanks so much. The hook.stopPropgation solved it for me.

Is there also a way to easily extract the filters from a request? For example:

`http://127.0.0.1:8090/api/collections/reviews/records?sort=-creation_date&filter=(product_url='https://ecommerce.com/product-1')`

Would it be possible to retrieve the value for filter "product_url" somewhere?

**Reply by ganigeorgiev** [reply]
Query parameters could be accessed by doing something like `e.HttpContext.QueryParam("filter")`, but the filter expression is not so easy to parse because technically it can have nested parenthesis, multiple condition join operators like `&&` and `||`, field modifiers, etc. and in this particular case it maybe easier to just extract the value with regex, eg. `product_url\s*=\s*(?:'|")(.+)(?:'|")`. Alternatively, if this is the only filter argument you want to access, you can also just send it as an additional query parameter and read it directly, eg. `?product_url=http...`.

**Reply by elmarw** [reply]
Yeah was considering regex.  Did not know about the single filter argument. That will work in my case. Cheers!

**Reply by ganigeorgiev** [reply]
> Did not know about the single filter argument. 

_Note that you still have to send the `filter` query param if you want the filter to be applied to the query. The suggestion with the **additional** query param was just an alternative way to avoid the regex if you don't want to deal with it._

**Reply by elmarw** [reply]
Ah I see, makes sense. Will go with regex for now.

**Reply by ganigeorgiev** [reply]
Yes, the regex is OK (at least for this particular use case). Here is one example how to read the captured group:

```go
re := regexp.MustCompile(`product_url\s*=\s*(?:'|")(.+)(?:'|")`)
match := re.FindStringSubmatch(e.HttpContext.QueryParam("filter"))
log.Println(match[1]) // https://ecommerce.com/product-1
```

**Reply by elmarw** [reply]
Ah wow, even more help than expected. I was just making something similar but you are way faster ;). Thanks!
